keps/sig-storage/20191213-csi-inline-volume-size.md

```diff
@@ -0,0 +1,175 @@
+for [CSI ephemeral inline
+volumes](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+currently supports only two standardized parameters for the volume:
+- `readOnly`
+- `fsType`
+
+All other parameters for the volume must be passed as
+`volumeAttributes` with key/value pairs that are driver-specific and
+thus opaque to Kubernetes. This API was chosen because it directly
+maps to parameters supported by the underlying
+[NodePublishVolumeRequest](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L3604-L3652).
+
+One common questions that developers and users have is how the size of
+the volume can be specified. The proposal is to introduce a
+standardized `fsSize` parameter as answer to that and gradually
+replace driver-specific parameters that may have been used before.
+
+## Motivation
+
+### Goals
+
+- Add a `CSIVolumeSource.fsSize` field.
+- Pass it to drivers as additional entry in `NodePublishVolumeRequest.VolumeContext`
+  if the driver has opted into getting additional fields there.
+
+### Non-Goals
+
+- Validate whether a CSI driver really supports this field. This is
+  consistent with `readOnly` and `fsType` which also may be silently
+  ignored.
+
+## Proposal
+
+### User Stories
+
+#### Ephemeral local volumes
+
+Local volumes are useful as scratch
+space. [PMEM-CSI](https://github.com/intel/pmem-csi) and
+[TopoLVM](https://github.com/cybozu-go/topolvm/blob/master/README.md)
+are two examples for CSI drivers which dynamically create volumes of a
+fixed capacity and thus need a size parameter.
+
+### Capacity-aware scheduling of pods
+
+Scheduling pods with ephemeral inline volumes onto nodes [with
+sufficient free storage
+capacity](https://github.com/kubernetes/enhancements/pull/1353)
+depends (among other information, like free capacity) on knowing the
+size of the volumes. If the size is only specified via some
+vendor-specific parameter, it's not available to the Kubernetes
+scheduler.
+
+### fsSize field
+
+A new field `fsSize` of type `*Quantity` in
+[CSIVolumeSource](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csivolumesource-v1-core)
+needs to be added, alongside the existing `fsType`. It must be a
+pointer to distinguish between no size and zero size selected.
+
+The new field is under the `CSIInlineVolumeSize` feature
+gate.
+
+While `fsType` can (and does get) mapped to
+`NodePublishVolumeRequest.volume_capability`, for `fsSize` we need a
+different approach for passing it to the CSI driver because there is
+no pre-defined field for it in `NodePublishVolumeRequest`. We can
+extend the [pod info on
+mount](https://kubernetes-csi.github.io/docs/pod-info.html) feature:
+if (and only if) the driver enables that, then a new
+`csi.storage.k8s.io/size` entry in
+`NodePublishVolumeRequest.publish_context` is set to the string
```

Comment:
It's not publish_context, but volume_context. For more information, please refer to the following PR.

https://github.com/kubernetes-csi/docs/pull/249

As a side note, I found this documentation error when reviewing this KEP, thanks!

Reply:
Good catch and thanks for taking care of it also in the docs from which I copied this. I wonder how many other developers have skipped over this without noticing the mistake...

Fixed.